src/core/Akka/Util/Internal/AtomicState.cs

```diff
@@ -0,0 +1,186 @@
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading.Tasks;
+
+namespace Akka.Util.Internal
+{
+    /// <summary>
+    /// Internal state abstraction
+    /// </summary>
+    internal abstract class AtomicState : AtomicCounterLong, IAtomicState
+    {
+        private readonly ConcurrentQueue<Action> _listeners;
+        private readonly TimeSpan _callTimeout;
+
+        protected AtomicState( TimeSpan callTimeout, long startingCount )
+            : base( startingCount )
+        {
+            _listeners = new ConcurrentQueue<Action>( );
+            _callTimeout = callTimeout;
+        }
+
+        /// <summary>
+        /// Add a listener function which is invoked on state entry
+        /// </summary>
+        /// <param name="listener">listener implementation</param>
+        public void AddListener( Action listener )
+        {
+            _listeners.Enqueue( listener );
+        }
+
+        /// <summary>
+        /// Test for whether listeners exist
+        /// </summary>
+        public bool HasListeners
+        {
+            get { return !_listeners.IsEmpty; }
+        }
+
+        /// <summary>
+        /// Notifies the listeners of the transition event via a 
+        /// </summary>
+        protected void NotifyTransitionListeners( )
+        {
+            if ( !HasListeners ) return;
+            Task
+                .Factory
+                .StartNew
+                (
+                    ( ) =>
+                    {
+                        foreach ( var listener in _listeners )
+                        {
+                            listener.Invoke( );
+                        }
+                    }
+                );
+        }
+
+        /// <summary>
+        /// Shared implementation of call across all states.  Thrown exception or execution of the call beyond the allowed
+        /// call timeout is counted as a failed call, otherwise a successful call
+        /// 
+        /// NOTE: In .Net there is no way to cancel an uncancellable task. We are merely cancelling the wait and marking this
+        /// as a failure.
+        /// 
+        /// see http://blogs.msdn.com/b/pfxteam/archive/2011/11/10/10235834.aspx 
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="task">Implementation of the call</param>
+        /// <returns>result of the call</returns>
+        public async Task<T> CallThrough<T>( Task<T> task )
+        {
+            var deadline = DateTime.Now.Add( _callTimeout );
+            try
+            {
+                return await task;
+            }
+            finally
+            {
+                switch ( task.Status )
+                {
+                    case TaskStatus.RanToCompletion:
+                        if ( DateTime.Now.CompareTo( deadline ) < 0 )
```

Comment:
See this thread on datetime.now https://github.com/akkadotnet/akka.net/issues/846